File: 2021-SOPM/E4/e04_1/app/src/main/java/com/example/e04_1/ContentActivity.java
package com.example.e04_1;

import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.e04_1.adapters.EmployeesAdapter;
import com.example.e04_1.enums.Gender;
import com.example.e04_1.enums.Role;
import com.example.e04_1.viewmodels.EmployeeViewModel;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.List;

public class ContentActivity extends AppCompatActivity {

    private final String TAG = "ContentActivity";
    private List<EmployeeViewModel> employeeViewModels = new ArrayList<>();
    private FirebaseDatabase firebaseDatabase;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        EmployeesAdapter employeesAdapter = new EmployeesAdapter();
        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setAdapter(employeesAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(view -> {
            EmployeeViewModel firstEmployee = new EmployeeViewModel("Armend", "Riinvest", "", Gender.Male, Role.Technology);
//            employeeViewModels.add(firstEmployee);
//            employeesAdapter.setData(employeeViewModels);
            firebaseDatabase = FirebaseDatabase.getInstance("https://example04-360e1-default-rtdb.europe-west1.firebasedatabase.app/");
            DatabaseReference ref = firebaseDatabase.getReference("users");

            String key = ref.push().getKey();
            ref.child(key).setValue(firstEmployee);
        });
    }
}

File: 2023-SOPM/ClassStudentManagement/app/src/main/java/com/example/classstudentmanagement/ContentActivity.kt
package com.example.classstudentmanagement

import android.os.Bundle
import android.util.Log
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import com.example.classstudentmanagement.adapters.CoursesAdapter
import com.example.classstudentmanagement.adapters.StudentAdapter
import com.example.classstudentmanagement.application.ClassStudentManagementApplication
import com.example.classstudentmanagement.models.Course
import com.example.classstudentmanagement.models.Student
import com.example.classstudentmanagement.services.ApiService
import com.example.classstudentmanagement.viewmodels.StudentViewModel
import com.example.classstudentmanagement.viewmodels.StudentViewModelFactory
import com.google.android.material.floatingactionbutton.FloatingActionButton
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ContentActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "ContentActivity"
        private var latestId = 5 // TODO: Don't use hard-coded number of Primary Keys
    }

    private val studentViewModel: StudentViewModel by viewModels {
        StudentViewModelFactory((application as ClassStudentManagementApplication).repository)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_content)

        val mainActivityIntent = intent

        val username = mainActivityIntent.getStringExtra("USERNAME")
        Log.d(TAG, "Username is ${username.toString()}")

        val recyclerView = createRecyclerView()

        // FAB onClick to add to DB
        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener {
            addNewStudentBasedOnLatestId()
        }

        callServiceToGetCoursesData(recyclerView!!)
    }

    private fun createRecyclerView(): RecyclerView? {
        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
        val studentAdapter = StudentAdapter()
        recyclerView.adapter = studentAdapter

        studentViewModel.students.observe(this, Observer { students ->
            students.let {
                studentAdapter.submitList(students)
                Log.d(TAG, "STUDENT Name $it")
            }
        })
        return recyclerView
    }

    private fun addNewStudentBasedOnLatestId() {
        latestId += 1
        val studentToAdd =
            Student(
                latestId,
                "Name$latestId",
                "surname$latestId",
                ""
            )
        studentViewModel.insert(studentToAdd)
    }

    private fun callServiceToGetCoursesData(recyclerView: RecyclerView) {
        // Calling a Service to get Courses Data
        val service = ApiService.getService().getCourses()

        service.enqueue(object : Callback<List<Course>> {
            override fun onResponse(call: Call<List<Course>>, response: Response<List<Course>>) {
                Log.d(TAG, "Response success: " + response.body().toString())

                if (response.isSuccessful) {
                    recyclerView.adapter = CoursesAdapter(response.body()!!)
                }
            }

            override fun onFailure(call: Call<List<Course>>, t: Throwable) {
                Log.d(TAG, "Response failed: " + t.message.toString())
            }
        })
    }
}